From sam@server1.gnuconsulting.com Thu Dec 22 16:15:04 2005
Date: Thu, 22 Dec 2005 17:11:02 -0700
From: Sam Bishop <sam@bishop.dhs.org>
To: Greg KH <greg@kroah.com>
Cc: <sbishop@micron.com>
Message-ID: <20051223001102.GA19026@server1.gnuconsulting.com>
Content-Disposition: inline
Subject: USB: fix usb-skeleton limit resource usage patch.

From: Sam Bishop <sam@bishop.dhs.org>

Prevents a compiler warning and uses down_interruptible() instead of down() in
process context.

Signed-off-by: Sam Bishop <sam@bishop.dhs.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/usb-skeleton.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- gregkh-2.6.orig/drivers/usb/usb-skeleton.c
+++ gregkh-2.6/drivers/usb/usb-skeleton.c
@@ -166,7 +166,7 @@ static ssize_t skel_write(struct file *f
 	int retval = 0;
 	struct urb *urb = NULL;
 	char *buf = NULL;
-	size_t writesize = min(count, MAX_TRANSFER);
+	size_t writesize = min(count, (size_t)MAX_TRANSFER);
 
 	dev = (struct usb_skel *)file->private_data;
 
@@ -175,7 +175,10 @@ static ssize_t skel_write(struct file *f
 		goto exit;
 
 	/* limit the number of URBs in flight to stop a user from using up all RAM */
-	down (&dev->limit_sem);
+	if (down_interruptible(&dev->limit_sem)) {
+		retval = -ERESTARTSYS;
+		goto exit;
+	}
 
 	/* create a urb, and a buffer for it, and copy the data to the urb */
 	urb = usb_alloc_urb(0, GFP_KERNEL);
